### PR TITLE
Drop messages instead of panicking when the debug buffer is full (Fix #1442).

### LINKED ISSUE #1442: Kernel panics when the debug buffer is full
Kernel debugging uses a circular buffer to store messages until they are sinked to a suitable output (UART, RTT, etc.). However, if this buffer gets full, the kernel immediately panics:

https://github.com/tock/tock/blob/95c1dc9d2665d4570afa26ab6935b0ade7fb88da/kernel/src/debug.rs#L485-L495

This is quite annoying when debugging protocols such as USB, where a lot of events happen quite frequently. This can be somewhat mitigated with a larger buffer (https://github.com/tock/tock/issues/1373) but this is still unreliable (e.g. sometimes the debug sink has a random delay and the kernel fills the buffer in the meantime).

Is there any reason for the kernel to panic when the circular buffer is full? I think that the debug messages should be dropped instead (possibly with a marker message such as `\n*** debug buffer full ***\n`), until the buffer gets flushed again.

### PR DESCRIPTION
**Pull request by gendx**
### Pull Request Overview

This pull request simplifies the implementation of kernel debugging, and avoids panicking when the debug buffer is full.
Instead, messages are dropped and a marker message is written to the debug output.

The implementation now relies on the kernel's `RingBuffer` type, which makes it much simpler.

### Testing Strategy

This pull request was tested by flashing Tock on a nRF52840dk board and writing large messages to the debug output. Also tested with a smaller `INTERNAL_BUF` array in order to trigger a full buffer more often.

### TODO or Help Wanted

1. For now, the implementation copies bytes one at a time. This may be slower than the previous implementation, but I think that correctness is more important. However, the `RingBuffer` type could be extended to enqueue/dequeue many elements at a time (i.e. via slices), to potentially allow faster copying.
1. Should a _component_ (https://github.com/tock/tock/pull/1338) be defined to initialize UART on boards?

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
👍 

This also really highlights that we should switch to components and make the updating work for changes like this a lot easier. I noticed some oddities (hail has the debug writer set three times...) so https://github.com/tock/tock/tree/pr/1448 is a branch that straightens out the components a little. That could be added to this PR, or we can just keep it separate and add it later.

File: kernel/src/common/ring_buffer.rs [hunk 1]
```diff
@@ -16,6 +16,31 @@ impl<T: Copy> RingBuffer<'a, T> {
...
     }
+
+    // Returns the number of elements that can be enqueued until the ring buffer is full.
+    pub fn available_len(&self) -> usize {
+        // The maximum capacity of the queue is ring.len - 1, because head == tail for the empty
+        // queue.
+        self.ring.len().saturating_sub(1 + queue::Queue::len(self))
+    }
+
+    pub fn as_slices(&'a self) -> (Option<&'a [T]>, Option<&'a [T]>) {
```
Comment (alevy) [inline kernel/src/common/ring_buffer.rs]: Please document.

File: kernel/src/common/ring_buffer.rs [hunk 2]
```diff
@@ -16,6 +16,31 @@ impl<T: Copy> RingBuffer<'a, T> {
             ring: ring,
         }
     }
+
+    // Returns the number of elements that can be enqueued until the ring buffer is full.
```
Comment (alevy) [inline kernel/src/common/ring_buffer.rs]: Three slashes for public documentation

File: kernel/src/debug.rs [hunk 3]
```diff
@@ -271,118 +264,50 @@ impl DebugWriter {
...
-
-            self.internal_buffer.map(|internal_buffer| {
-                for (dst, src) in out_buffer
-                    .iter_mut()
-                    .zip(internal_buffer[start..real_end].iter())
-                {
-                    *dst = *src;
+            let mut count = 0;
+
+            let ring_buffer = self.internal_buffer.take().unwrap();
```
Comment (ppannuto) [inline kernel/src/debug.rs]: `unwrap` always makes me nervous, but I'm not sure what the behavior here should be if the buffer isn't available?

Comment (gendx) [reply kernel/src/debug.rs]: Good point! I guess I should use `.map()` instead of `.take()`. Will try that & fix the other comments.

**Comment by gendx** [comment]
Your comments should be resolved now.

**Comment by phil-levis** [comment]
I'm too late to the party!

This seems like a good analogue to a discussion Brad and I had about how to handle serious errors in the kernel. There are absolutely cases when I want debug output to panic if I overflow, in which I want this it to fail fast, so that I don't later find out that actually I was losing the messages I really needed. That being said, I agree this should not be the default behavior.

**Pull request merged**